### ISSUE TITLE
Create Live Stream scraper

### ISSUE THREAD

**Issue by caffeinatedMike** [caffeinatedMike/plugin.video.Squeee#3]
**Url for original Http request to get unresolved links:**
[http://www.ustream.tv/search?q=animals&category=pets-animals&type=live&location=anywhere](http://www.ustream.tv/search?q=animals&category=pets-animals&type=live&location=anywhere)

**Needed content of original request can be found in the following tags:**
```
<div class="item media-item">
	<div class="item-image">
		<a href="/channel/apl-puppies" title="APL!VE Puppies">
			<img src="https://ustvstaticcdn1-a.akamaihd.net/i/explore/custom/1/3/0/2/13027422/13027422_channel_en_US_wide,192x108,r:1360182679.jpg"
<p class="item-channel-tbn">
		</a>
	</div>
	...
</div>
```

**Api Http request needed for subsequent calls to obtain actual streaming links:**
[https://api.ustream.tv/channels/{channel_id}.json](https://api.ustream.tv/channels/12623552.json)

**Needed content of subsequent calls can be found in the following json fields:**
```
"status": "live"
"thumbnail":{"live":"https://ustvstaticcdn1-a.akamaihd.net/i/channel/live/1_12623552,192x108,b:2017061409.jpg"}
"stream":{"hls":"http://iphone-streaming.ustream.tv/uhls/12623552/streams/live/iphone/playlist.m3u8"}
```

Reference [HOW-TO play UStream livestreams](https://forum.kodi.tv/showthread.php?tid=109475) for a kodi-friendly method if above does not work.

### Streamlink also already has a [kodi plugin](https://github.com/beardypig/plugin.video.streamlink), which can prove useful for further methodology.

**Comment by caffeinatedMike** [comment]
The following proof of concept (credit to @beardypig for providing framework) works to play original stream segment, but routine polling is needed to keep stream alive.

```
def get_uhs_client():
    client = UHSClient(11697871, 'channel')
    if client.connect():
        if client.poll() is None:
            if client.connect():
                stream_url = client.poll()
                print stream_url
                play_item = xbmcgui.ListItem(path=stream_url)
                play_item.setProperty('IsPlayable','true')
                xbmcplugin.setResolvedUrl(_handle, True, listitem=play_item)
        else:
            print client.poll()

class UHSClient(object):
    """
    API Client, reverse engineered by observing the interactions
    between the web browser and the ustream servers.
    """
    API_URL = "http://r{0}-1-{1}-{2}-{3}.ums.ustream.tv"
    APP_ID, APP_VERSION = 11, 2
    USER_AGENT = ("Mozilla/5.0 (iPhone; CPU iPhone OS 6_0 like Mac OS X) "
                  "AppleWebKit/536.26 (KHTML, like Gecko) Version/6.0 "
                  "Mobile/10A5376e Safari/8536.25")

    def __init__(self, media_id, application, **options):
        self.media_id = media_id
        self.application = application
        self.referrer = options.pop("referrer", None)
        self._host = None
        self.rsid = self.generate_rsid()
        self.rpin = self.generate_rpin()
        self._connection_id = None
        self._app_id = options.pop("app_id", self.APP_ID)
        self._app_version = options.pop("app_version", self.APP_VERSION)
        self._cluster = options.pop("cluster", "live")
        self._password = options.pop("password", None)

    def connect(self, **options):
        raw = self.send_command(type="viewer", appId=self._app_id,
                                   appVersion=self._app_version,
                                   rsid=self.rsid,
                                   rpin=self.rpin,
                                   referrer=self.referrer,
                                   media=str(self.media_id),
                                   application=self.application,
                                   password=self._password)
        result = raw[0]["args"][0]
        self._host = "http://{0}".format(result["host"])
        self._connection_id = result["connectionId"]
        log.debug("Got new host={0}, and connectionId={1}", self._host, self._connection_id)
        return True

    def poll(self, schema=None, retries=5, timeout=5.0):
        stime = time.time()
        try:
            r = self.send_command(connectionId=self._connection_id,
                                  schema=schema,
                                  retries=retries,
                                  timeout=timeout)
            if r[0]['cmd'] == 'reject':
                print 'rejected'
                self.referrer = r[0]['args'][0]['referrerLock']['redirectUrl']
                print self.referrer
                connect()
            return r[0]['args'][0]['stream'][0]['url']
        except Exception as err:
            log.debug("poll took {0:.2f}s: {1}", time.time() - stime, err)
        else:
            log.debug("poll took {0:.2f}s", time.time() - stime)
            return r

    @staticmethod
    def generate_rsid():
        return "{0:x}:{1:x}".format(randint(0, 1e10), randint(0, 1e10))

    @staticmethod
    def generate_rpin():
        return "_rpin.{0}".format(randint(0, 1e15))

    def send_command(self, timeout=5.0, **args):
        res = requests.get(self.host,
                           params=args,
                           headers={"Referer": self.referrer,
                                    "User-Agent": self.USER_AGENT},
                           timeout=timeout)
        return json.loads(res.text)

    @property
    def host(self):
        host = self._host or self.API_URL.format(randint(0, 0xffffff), self.media_id, self.application,
                                                 "lp-" + self._cluster)
        return urljoin(host, "/1/ustream")
```

**Issue/TODO:**
- [ ] Implement Polling
- [ ] Implement Proxy Service

**Comment by caffeinatedMike** [comment]
Rough pseudo-code for polling service
```
import xbmc
from resources import liveclient

# Get the plugin url in plugin:// notation.
_url = sys.argv[0]
# Get the plugin handle as an integer number.
_handle = int(sys.argv[1])

class LiveMonitor(xbmc.Monitor)
    def __init__(self):
        xbmc.Monitor.__init__(self)
        self.Player = LivePlayer()
    def OnSettingsChange(self):
        live = addon.getSetting('live_url')
        if live != '':
            client = liveclient.UHSClient(live, 'channel')
            client.poll()
        else:
            pass
class LivePlayer(xbmc.Player)
    def __init__(self):
        xbmc.Player.__init__(self)
    def onPlayBackStopped():
        addon.setSetting('live_url') = ''

if __name__ == "__main__":

    monitor = LiveMonitor()
    
    # XBMC loop
    while not monitor.waitForAbort(10):
        pass
```

**Comment by caffeinatedMike** [comment]
The posts referenced in the original issue comment are for ustream's old api. The above polling service is not needed as a separate service, thus eliminating the need to extend . Instead, the proof of concept provided above can be applied with a built-in Player subclass (again, thanks & credit goes to @beardypig) as is implemented in [this gist](https://gist.github.com/beardypig/4e81c76a34137b2ef0d53625525f7895) with the following invocation
```
UStreamPlayer().playChannel(11697871)
```

Moving forward using this implementation as the basis, the following other obstacles & questions need to be addressed:

1. How to get data from HTTP GET. When visiting the [search url](http://www.ustream.tv/search?q=animals&category=pets-animals&type=live&location=anywhere) in a web browser & inspecting the page the needed tags can be easily identified. However, when performing a request from a REST client or via code provides different results with the relevant data not even being displayed in the response.

**Answer:** Through some digging in the page's script tags I found the api search url that provides a json response. Using [this url](http://www.ustream.tv/ajax/search.json?q=) provides the html code of the actual page in the `pageContent` field.

2. Which search(es) to use. Most searches tried heed very limited results (2 pages). Should we
 - Create a hand-picked list of streams to list & provide a Search feature in case the user wants to see if there is something else not specifically listed?
 - Use a select one or few of the searches and simply display those results?
To date, I've tested the following searches:
 - `q=animals&type=live&category=pets-animals` (2 pages of results)
 - `q=APL!VE&type=live&category=pets-animals` (1 page of 4 results)
 - `q=animals&type=live` (2 pages of results) _Note:_ Contained at least 1 unrelated stream

In my opinion, I believe that hand-picking a select variety of streams & allowing a search function using the query `http://www.ustream.tv/search?q={search term}&category=pets-animals&type=live` would heed the best results for quality and quantity of streams.
